chore: add docs for new filters functionality (#6458)
obot-platform/obot: docs/docs/functionality/filters.md

Reviewer: nit: Maybe add something here like "All filter types support..."

Author: Done.

Filters are a powerful mechanism for inspecting and controlling tool calls and their results in the MCP Gateway. They provide administrators with the ability to implement custom validation, logging, security checks, or other business logic by intercepting tool requests and responses before they are processed.

Filters can be implemented in two ways:

- **MCP filter servers**: MCP servers that expose a filter tool. Obot deploys and calls the configured tool when matching MCP traffic is processed.
- **HTTP webhook filters**: HTTP endpoints that receive MCP messages from the gateway.

When you configure a filter, you can narrow when it runs using selectors that target particular tool calls or MCP (Model Context Protocol) tool functions.

## How Filters Work

> 1. **Tool Call Interception**: When a tool call is made, the gateway intercepts it and sends the details to your configured filter

Reviewer 2: If I understand correctly, any type of MCP message can be filtered, right? Not just tool calls?

Author: Adjusted.

Filters are a powerful mechanism for inspecting and controlling tool calls and their results in the MCP Gateway. They provide administrators with the ability to implement custom validation, logging, security checks, or other business logic by intercepting tool requests and responses before they are processed.

Filters can be implemented in two ways:

- **MCP filter servers**: MCP servers that expose a filter tool. Obot deploys and calls the configured tool when matching MCP traffic is processed.
- **HTTP webhook filters**: HTTP endpoints that receive MCP messages from the gateway.

When you configure a filter, you can narrow when it runs using selectors that target particular tool calls or MCP (Model Context Protocol) tool functions.

## How Filters Work

> 1. **MCP Request Interception**: When a tool call is made, the gateway intercepts it and sends the details to your configured filter

Reviewer 2: ```suggestion
1. **MCP Request Interception**: When an MCP request is made, the gateway intercepts it and sends the details to your configured filter
```

I think the wording here could still be improved a bit

> 1. **MCP Request Interception**: When a is made to an MCP server, the gateway intercepts it and sends the details to your configured filter

Reviewer 2: ```suggestion
1. **MCP Request Interception**: When a request is made to an MCP server, the gateway intercepts it and sends the details to your configured filter
```

Author: I'll blame this on one the learning disability.

The additional requirement for an MCP filter server is a filter tool name. Obot needs this value so it knows which tool to call when the filter runs.

### Filter Tool Contract

When you write an MCP server that acts as a filter, implement the tool called by `filterConfig.toolName` with this contract:

1. Obot calls one tool on the MCP server for filtering.
2. The tool accepts the full MCP message as an argument.
3. The tool returns a response with these fields:

| Field | Description |
|-------|-------------|
| `accept` | Whether the MCP message should be allowed to continue. |
| `mutated` | Whether the filter returned a modified MCP message. Use this only when the filter is allowed to mutate traffic. |
| `message` | The MCP message to use after filtering. For non-mutating filters, return the original message. |
| `reason` | Human-readable explanation for the decision. Include this when rejecting or mutating a message so administrators can understand the result. |

See the [obot-platform/pii-filter](https://github.com/obot-platform/pii-filter) repository for an example MCP filter server.

## Built-in Filters

> Obot also supports built-in filters. These are MCP filter servers that are already configured for deployment in Obot through the system MCP catalog.

Reviewer: nit: Alternatively start this with `Obot ships with a default set of built-in filters...`

Author: Done.
